### ISSUE TITLE
val() and predict() produce inconsistent prediction results

### ISSUE THREAD

**Issue by cavalier501** [ultralytics/ultralytics#22088]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello,

I am using a detection model and have been running the `val()` and `predict() `functions on the same checkpoint `(ckpt)`. Both functions are applied to the same image and the results are visualised. However, the outputs from the two functions are inconsistent. I would like to know if there is an issue with my setup.

Parameter settings for the two functions val():

```
metrics = model.val(
    data=eval_dataset["yaml_path"],
    imgsz=640,
    batch=16,
    device=[0,],
    workers=8,
    project=project_name,
    single_cls=True,
    visualize=True,
    # conf and iou should be default 0.001 and 0.7
) 
```

predict()

```
model.predict(
    source=tmp_dir, 
    conf=0.25, 
    iou=0.7, 
    imgsz=640,
    batch=16,
    device=[0],
    max_det=300, 
    classes=None, 
    save=True,
    project=pred_project_dir,
    name=None,
    save_txt=True, 
    save_conf=True, 
)
```

In some case, the `predict()` result can predict box with conf 0.4 while `val()` can't predict box in the same position of same image, in some case, the `val() `result shows two bounding boxes in the image with conf 0.5 and 0.5, whereas the `predict()` result shows only one with conf 0.37, 

Additionally, how can I modify the visualizations of `val()` and `predict()` so that the confidence shown on the predicted boxes is displayed with a specified number of decimal places?

Thank you for your help.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @cavalier501, thanks for reporting this and for using Ultralytics 🚀 This is an automated response to help get things moving quickly—an Ultralytics engineer will review and assist you shortly.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. Since you’re seeing differing results between `val()` and `predict()`, please include:
- Exact `ultralytics` version and environment details (OS, CUDA/CUDNN, Python, PyTorch)
- The precise commands/snippets you ran for both `val()` and `predict()`
- One or two sample images and the exact checkpoint used
- All inference parameters that may differ (e.g., `conf`, `iou`, `single_cls`, `classes`, `max_det`, `batch`, `imgsz`, any TTA/augment settings)
- Whether any preprocessing/visualization steps differ between runs
- Console logs for both runs with any warnings shown

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by cavalier501** [comment]
Additionally, how can I modify the visualizations of `val()` and `predict()` so that the confidence shown on the predicted boxes is displayed with a specified number of decimal places?

**Comment by Y-T-G** [comment]
Duplicate of https://github.com/ultralytics/ultralytics/issues/2783

**Comment by Y-T-G** [comment]
You can read this

https://github.com/ultralytics/ultralytics/issues/2783#issuecomment-2413530537

**Comment by Y-T-G** [comment]
> Additionally, how can I modify the visualizations of val() and predict() so that the confidence shown on the predicted boxes is displayed with a specified number of decimal places?

You can update `.2f` to any other value you want

https://github.com/ultralytics/ultralytics/blob/f0c7e7b3c95d22655f9ddc216bd954c3f7007b1f/ultralytics/engine/results.py#L571

**Comment by cavalier501** [comment]
> You can read this
> 
> [#2783 (comment)](https://github.com/ultralytics/ultralytics/issues/2783#issuecomment-2413530537)

Thanks for your reply. But I have some additional questions regarding this. 

I have understood that `val()` and `predict()` behave differently with respect to image resolution.  
But how can I obtain the `rect=True` result in `predict()` mode?  
No matter whether I set `rect=True` or `rect=False` in `predict()`, I cannot achieve it.

In the issue you mentioned, I see that the padding behavior is consistent between `val()` and `predict()` (`pad=0.5`), whereas it differs from `train()` (`pad=0`).  
Are there any previous experiments that indicate which padding setting is more advantageous for better performance?

**Comment by Y-T-G** [comment]
Padding here is just increasing the image size. So 640 would become 672. You can try running `predict()` with `imgsz=672`. But it won't exactly match `val()` for other reasons like `val()` using multi-label prediction and `val()` running in batches which calculates minimum rectangle on the whole batch, and not a single image.

**Comment by Y-T-G** [comment]
Or you can run both `predict()` and `val()` with `rect=False` to get similar results.

**Comment by cavalier501** [comment]
> Padding here is just increasing the image size. So 640 would become 672. You can try running `predict()` with `imgsz=672`. But it won't exactly match `val()` for other reasons like `val()` using multi-label prediction and `val()` running in batches which calculates minimum rectangle on the whole batch, and not a single image.

Thank you for your reply, and sorry for the delayed response.

I have another question to ask: In val() with save_txt=True, how should the TXT label files in val/labels be interpreted? I noticed they don't seem to follow the normalized "cx, cy, w, h" format — some coordinates are greater than 1, and if interpreted as "cx, cy, w, h", their positions do not match the corresponding visualizations in val/visualizations.

Could you please clarify the format of these label files? Also, how can I obtain the standard normalized "cx, cy, w, h" format labels for validation?

**Comment by Y-T-G** [comment]
They are in `cx cy w h` format.

**Comment by cavalier501** [comment]
> They are in `cx cy w h` format.

Uh, I found one txt file under val/labels containing the following result:

```
0 1.01879 1.01386 0.379273 0.382178 0.00482224

0 0.625373 0.769689 0.950556 0.880252 0.00268142

0 0.948246 1.00463 0.523859 0.39701 0.00201886
```

According to the (cx, cy, w, h) format, I understand that coordinates should not be greater than 1.

Moreover, when I compared these values with those in the corresponding visualization results under val/visualizations, the coordinates in the txt files don’t match exactly. If I need some transformation to obtain the actual (cx, cy, w, h) coordinates?

**Comment by glenn-jocher** [comment]
Thanks for the examples—this is expected. Files under `val/labels` are YOLO‑normalized `[class x_center y_center width height]` with an optional trailing confidence when `save_conf=True`; values may be slightly >1 or <0 if a box extends past image bounds (we don’t clamp before writing), while the visualizations clip to the image, so they won’t match exactly character‑for‑character. If you want strictly bounded [0,1] without confidences, re‑run validation with:
```python
metrics = model.val(data=..., imgsz=640, save_txt=True, save_conf=False)
```
Format/args are documented in the [validation mode guide](https://docs.ultralytics.com/modes/val/) and the save format is defined by `Results.save_txt()` using normalized `xywhn` coordinates as described in the [Results reference](https://docs.ultralytics.com/reference/engine/results/#save_txt) and its [`xywhn` property](https://docs.ultralytics.com/reference/engine/results/#xywhn-property).

**Comment by cavalier501** [comment]
> Thanks for the examples—this is expected. Files under `val/labels` are YOLO‑normalized `[class x_center y_center width height]` with an optional trailing confidence when `save_conf=True`; values may be slightly >1 or <0 if a box extends past image bounds (we don’t clamp before writing), while the visualizations clip to the image, so they won’t match exactly character‑for‑character. If you want strictly bounded [0,1] without confidences, re‑run validation with:
> 
> metrics = model.val(data=..., imgsz=640, save_txt=True, save_conf=False)
> Format/args are documented in the [validation mode guide](https://docs.ultralytics.com/modes/val/) and the save format is defined by `Results.save_txt()` using normalized `xywhn` coordinates as described in the [Results reference](https://docs.ultralytics.com/reference/engine/results/#save_txt) and its [`xywhn` property](https://docs.ultralytics.com/reference/engine/results/#xywhn-property).

Thanks much for your reply. As I understand it, your reply means that, when the predicted x_center or y_center is greater than 1, it actually indicates that the model believes there is a detection located outside the image boundaries. Does this mean that such detections should actually be discarded?

Furthermore, could you please explain how these boxes are generated? Since YOLO’s predictions are anchor point‑based, I suspect this might be related to the padding performed in `eval()`. Does this imply that, in evaluation mode, the `d.xywhn` coordinates in the results needs an additional transformation to the detected box’s <x_center, y_center, width, height> on the actual image?

Additionally, I checked the implementation of `Results.save_txt()`, and it doesn’t seem to clamp the values to the [0,1] range as you suggested. Actually. even with `save_conf=False`, I still see x_center/y_center values that exceed 1.

The` Results.save_txt()` script are as follow:

```
is_obb = self.obb is not None
boxes = self.obb if is_obb else self.boxes
masks = self.masks
probs = self.probs
kpts = self.keypoints
texts = []
if probs is not None:
    [texts.append(f"{probs.data[j]:.2f} {self.names[j]}") for j in probs.top5]
elif boxes:
    # Detect/segment/pose
    for j, d in enumerate(boxes):
        c, conf, id = int(d.cls), float(d.conf), int(d.id.item()) if d.is_track else None
        line = (c, *(d.xyxyxyxyn.view(-1) if is_obb else d.xywhn.view(-1)))
        if masks:
            seg = masks[j].xyn[0].copy().reshape(-1)  # reversed mask.xyn, (n,2) to (n*2)
            line = (c, *seg)
        if kpts is not None:
            kpt = torch.cat((kpts[j].xyn, kpts[j].conf[..., None]), 2) if kpts[j].has_visible else kpts[j].xyn
            line += (*kpt.reshape(-1).tolist(),)
        # save_conf only take effect this line
        line += (conf,) * save_conf + (() if id is None else (id,))
        texts.append(("%g " * len(line)).rstrip() % line)

if texts:
    Path(txt_file).parent.mkdir(parents=True, exist_ok=True)  # make directory
    with open(txt_file, "a", encoding="utf-8") as f:
        f.writelines(text + "\n" for text in texts)

return str(txt_file)
```

**Comment by glenn-jocher** [comment]
Great questions—quick clarifications:

- Values >1 or <0 in `val/labels/*.txt` are expected when a predicted box lies partly outside the original image after reverse letterboxing; we don’t clamp before writing. No extra transform is required—`val()` already maps predictions back to the original image via `scale_preds()` and then writes normalized `cx cy w h [conf]` in `save_one_txt()`; see the validator’s `scale_preds()` and `save_one_txt()` references for the exact flow. For details on the TXT format used by `save_txt`, see the Predict settings table.  
  - `scale_preds()` reference: [Detect validator: scale_preds](https://docs.ultralytics.com/reference/models/yolo/detect/val/#scale_preds)  
  - `save_one_txt()` reference: [Detect validator: save_one_txt](https://docs.ultralytics.com/reference/models/yolo/detect/val/#save_one_txt)  
  - TXT format reference: [Predict settings: save_txt/save_conf](https://docs.ultralytics.com/usage/cfg/#predict-settings)

- Whether to discard these out‑of‑bounds boxes is up to your downstream use. If you need strict [0,1] normalization for post‑processing, clamp (and optionally drop zero‑area boxes) before saving. Example:

```python
from ultralytics import YOLO
import torch, numpy as np
from pathlib import Path

model = YOLO("yolo11n.pt")
out_dir = Path("labels_clamped"); out_dir.mkdir(parents=True, exist_ok=True)

for r in model.predict(source="your/images", stream=True, imgsz=640):
    xywhn = r.boxes.xywhn.clamp(0, 1)
    cls = r.boxes.cls.to(torch.int64).unsqueeze(1)
    conf = r.boxes.conf.unsqueeze(1)
    arr = torch.cat([cls, xywhn, conf], 1).cpu().numpy()
    np.savetxt(out_dir / (Path(r.path).stem + ".txt"),
               arr, fmt=["%d","%.6f","%.6f","%.6f","%.6f","%.6f"])
```

Note: you’re right—`save_conf=False` doesn’t clamp; my earlier wording was imprecise. If you prefer to clamp the files produced by `val(save_txt=True)`, load each TXT, clip columns 1:5 to [0,1], and re‑save.
